src/main/java/org/cryptomator/ui/mainwindow/MainWindowController.java

```diff
@@ -47,4 +55,37 @@ private void mainWindowFocusChanged(Observable observable) {
 		}
 	}
 
+	private void checkDisplayBounds(WindowEvent event) {
+		if (isOutOfDisplayBounds()) {
+			// If the position is illegal, then the window appears on the main screen in the middle of the window.
+			Rectangle2D primaryScreenBounds = Screen.getPrimary().getBounds();
+			window.setX((primaryScreenBounds.getWidth() - window.getMinWidth()) / 2);
+			window.setY((primaryScreenBounds.getHeight() - window.getMinHeight()) / 2);
+			window.setWidth(window.getMinWidth());
+			window.setHeight(window.getMinHeight());
+			savePositionalSettings();
+		}
+	}
+
+	private boolean isOutOfDisplayBounds() {
+		final double x = window.getX() + 1;
+		final double y = window.getY() + 1;
+		final double width = window.getWidth() - 2;
+		final double height = window.getHeight() - 2;
+		return isRectangleOutOfScreen(x, y, 0, height) // Left pixel column
+				|| isRectangleOutOfScreen(x + width, y, 0, height) // Right pixel column
+				|| isRectangleOutOfScreen(x, y, width, 0) // Top pixel row
+				|| isRectangleOutOfScreen(x, y + height, width, 0); // Bottom pixel row
+	}
+
+	private boolean isRectangleOutOfScreen(double x, double y, double width, double height) {
+		return Screen.getScreensForRectangle(x, y, width, height).isEmpty();
+	}
```

Comment:
The intend of `isOutOfDisplayBounds()` seems to be this: Consider a window out of bounds, if at least one of its edges is out of bounds. In other words: The window is out of bounds, if it is not completely within bounds.

This definition is debatable, but for our purpose I'm fine with it. For some fine tuning, however, consider the following example:

![Bounds Check 1](https://github.com/cryptomator/cryptomator/assets/1204330/5d0ab0d1-581b-4d30-a4f7-4952544dcee6)

All green edges are at least partially intersecting at least one screen, even if these edges are just a single pixel in width/height. Shouldn't this be "ouf of bounds", triggering a reset to the center? We could easily "broaden" the edges, needing at least 10px intersections with one of the screens.

Reply:
The situation in your example should indeed be out of bounds. I think I know another solution. Counting the total visible area of the window and comparing it to the actual area of the window. If the counted area does not match the actual area then it is not within display bounds. I will try it out and test if that works this week. And I agree that `isOutOfDisplayBounds()` might not be the best name. Maybe `isWithinDisplayBounds()` or `isPartiallyOutOfDisplayBounds()` is a better name.

Reply:
> The situation in your example should indeed be out of bounds.

As stated above, I'm actually fine with the previous definition. I'm just not sure if this was the intended way. In fact I like the previous implementation far better for its simplicity (and not needing any "double"-based math), however we should make sure to intersect more than just a single pixel column or row. In order to do that, we can inset the window rect by e.g. 10px and feed the four edges of the smaller rect into the "has partial intersection with at least one screen" function.

Remember: The issue we're trying to fix here is that the window is completely out of the screen area, so users can't see it after disconnecting a screen.

Reply:
So if I am not mistaken, you are saying that the case from your example should not be reset if the width/height of the visible part of the window is more than e.g. 10px?

I think that can easily be done by checking if there are > 0 screens for the blue rectangle. Here the blue rectangle is e.g. 10px smaller in all directions compared to the main window.
![img](https://github.com/cryptomator/cryptomator/assets/14298143/fb6d3178-f884-4a99-b90c-be248ca08d7a)

>The issue we're trying to fix here is that the window is completely out of the screen area, so users can't see it after disconnecting a screen.

This is true but there is also another issue: title bar out of display bounds, making it (near) impossible to move the window around. I thought it would be the easiest to just reset whenever a single pixel is out of bounds but if I understand you correctly you don't agree, which is fine :).

Also, I don't think it is very user friendly if the window is e.g. 90% out of bounds but is not reset because the visible width/height is more than e.g. 10 px. May I suggest using something like 50% of the width for horizontal direction and bottom direction. To make sure the title bar is visible we should do maybe 50% of the height of the title bar in the top of the screen. Percentages because not all screens have the same resolution. This results in something like this: If any of the blue areas are completely out of bounds then reset the window.
![suggested](https://github.com/cryptomator/cryptomator/assets/14298143/47ed087e-b4cb-46d7-9d55-12c070f6ad49)

Before I make any changes please let me know if I understand you correctly and if so, your thoughts about the percentage of the width and height that must be visible to trigger a reset.

Reply:
> May I suggest using something like 50% of the width for horizontal direction and bottom direction. To make sure the title bar is visible we should do maybe 50% of the height of the title bar in the top of the screen. Percentages because not all screens have the same resolution.

Yes, this is a reasonable suggestion, since not all areas are equally important when it comes to recovering a partially off-screen window.

We'll also have some offline-discussions about it and I'll chime back in here when we can precisely define the rects.